### DISCUSSION TITLE
图床转换工具-picvt

### DISCUSSION THREAD

**Discussion by caibingcheng** [caibingcheng/comment#118] [Announcements]
https://www.bbing.com.cn/202205/aboutpicvt/

**Comment by caibingcheng** [comment]
支持去不图床的upload，感谢[杜老师说](https://dusays.com/7bu/#atk-comment-5951)提供的帮助

**Comment by penndu** [comment]
哇哦，也支持去不图床啊！

**Comment by caibingcheng** [comment]
> @penndu
> 哇哦，也支持去不图床啊！

前天更新的, 自用, 没经过稳定性测试... 杜老师有什么类似工具推荐吗?

**Comment by penndu** [comment]
> @caibingcheng
> > @penndu
> > 哇哦，也支持去不图床啊！
> 
> 前天更新的, 自用, 没经过稳定性测试... 杜老师有什么类似工具推荐吗?

图片迁移？

**Comment by caibingcheng** [comment]
> @penndu
> > @caibingcheng
> > > @penndu
> > > 哇哦，也支持去不图床啊！
> > 
> > 前天更新的, 自用, 没经过稳定性测试... 杜老师有什么类似工具推荐吗?
> 
> 图片迁移？

是的, markdown图床切换一类

**Comment by penndu** [comment]
> @caibingcheng
> > @penndu
> > > @caibingcheng
> > > > @penndu
> > > > 哇哦，也支持去不图床啊！
> > > 
> > > 前天更新的, 自用, 没经过稳定性测试... 杜老师有什么类似工具推荐吗?
> > 
> > 图片迁移？
> 
> 是的, markdown图床切换一类

可以推荐大家试试！

**Comment by penndu** [comment]
> @penndu
> > @caibingcheng
> > > @penndu
> > > > @caibingcheng
> > > > > @penndu
> > > > > 哇哦，也支持去不图床啊！
> > > > 
> > > > 前天更新的, 自用, 没经过稳定性测试... 杜老师有什么类似工具推荐吗?
> > > 
> > > 图片迁移？
> > 
> > 是的, markdown图床切换一类
> 
> 可以推荐大家试试！

过两天发一个推文，又能水一篇了！

**Comment by GC-ZF** [comment]
你好冰冰😁，我最近也想迁移一下我的图到去不，所以杜老师推荐我来读这篇。但是代码貌似不能用了？首先，企业版兰空在前段时间进行了大更，token只能从面板申请，post请求不再适用需要手动申请，手动修改源码`class API7BU`中的`init_token`申请和`del_token`删除这俩函数后，代码跑通了，之后运行
```
python3 ./picvt.py -D ../blog/content/ -F local -T 7bu --path /home/ubuntu/play/blog/content/ --user xxx@qq.com --password xxx
```
可以跑通但是输出
```
Continue: [Y/n]Y
success 0/0
```
其中`/home/ubuntu/play/blog/content/`存放的是图片，如`/home/ubuntu/play/blog/content/xxx.png`与`/home/ubuntu/play/blog/content/xxx/xxx.png`，请问上传失败是因为填写的路径有问题么

**Comment by caibingcheng** [comment]
> @GC-ZF
> 你好冰冰😁，我最近也想迁移一下我的图到去不，所以杜老师推荐我来读这篇。但是代码貌似不能用了？首先，企业版兰空在前段时间进行了大更，token只能从面板申请，post请求不再适用需要手动申请，手动修改源码`class API7BU`中的`init_token`申请和`del_token`删除这俩函数后，代码跑通了，之后运行
> ```
> python3 ./picvt.py -D ../blog/content/ -F local -T 7bu --path /home/ubuntu/play/blog/content/ --user xxx@qq.com --password xxx
> ```
> 可以跑通但是输出
> ```
> Continue: [Y/n]Y
> success 0/0
> ```
> 其中`/home/ubuntu/play/blog/content/`存放的是图片，如`/home/ubuntu/play/blog/content/xxx.png`与`/home/ubuntu/play/blog/content/xxx/xxx.png`，请问上传失败是因为填写的路径有问题么

看你的参数是从本地迁移到去不吗？可以检查一下 `-D ../blog/content/` 这个设置对了吗？这个需要是你文章的md的目录，设置错了会找不到目标；`--path /home/ubuntu/play/blog/content/`看着是没问题的，因为你是本地存储，所以这里需要填写本地图片的存储路径。。。另外，感谢提醒， 我刚刚更新了一笔，推送到7bu删除了`--user`和`--password`参数，使用`--token`即可，可以试着更新一下

**Comment by GC-ZF** [comment]
原来如此，我把参数理解错误了，但还有点其它疑问
我的目录如下
```
.
├── blog
│   ├── cmd.md
│   └── img
│       └── cmd
│           ├── cmd01.png
│           ├── cmd02.png
│           ├── cmd03.png
│           ├── cmd04.png
│           └── cmd05.png
├── parser.py
├── picvt.py
├── platforms
│   ├── __init__.py
│   ├── picvt_7bu.py
│   ├── picvt_base.py
│   ├── picvt_config.json
│   ├── picvt_github.py
│   ├── picvt_imgtu.py
│   └── picvt_local.py
```
运行
```
python3 ./picvt.py -D /home/ubuntu/play/picvt/blog/ -F local -T 7bu --path /home/ubuntu/play/picvt/blog/ --user xxx --token atlbU9CTH
```
报错
```
/home/ubuntu/play/picvt/blog/cmd.md /img/cmd/cmd02.png)
/home/ubuntu/play/picvt/blog/cmd.md /img/cmd/cmd03.png)
/home/ubuntu/play/picvt/blog/cmd.md /img/cmd/cmd04.png)
/home/ubuntu/play/picvt/blog/cmd.md /img/cmd/cmd05.png)
Continue: [Y/n]Y
[1/4] fetching /img/cmd/cmd02.png) ok
[1/4] saving to /home/ubuntu/play/picvt/blog/img/cmd/cmd02.png)
[1/4] uploading Traceback (most recent call last):
  File "./picvt.py", line 82, in <module>
    upload_ok, url = params['to']['execute'].upload(
  File "/home/ubuntu/play/picvt/platforms/picvt_7bu.py", line 50, in upload
    'file': open(path, 'rb'),
FileNotFoundError: [Errno 2] No such file or directory: '/home/ubuntu/play/picvt/blog/img/cmd/cmd02.png)'
```
目录不存在？打扰前辈了，如果忙的话就算了，批量上传我自己做了一些实现了，杜老师看到后说让我来看看你的实现方案，所以前来研读一下

**Comment by caibingcheng** [comment]
> @GC-ZF

这里，路径名匹配错了`'/home/ubuntu/play/picvt/blog/img/cmd/cmd02.png)'`，多匹配了一个`)`
> FileNotFoundError: [Errno 2] No such file or directory: '/home/ubuntu/play/picvt/blog/img/cmd/cmd02.png)'

（PS：感觉问题好多，我只测试过自己遇到过的case，所以有好多情况可能不适配 😭 抱歉抱歉~~）

**Comment by GC-ZF** [comment]
> 这里，路径名匹配错了`'/home/ubuntu/play/picvt/blog/img/cmd/cmd02.png)'`，多匹配了一个`)`
> > FileNotFoundError: [Errno 2] No such file or directory: '/home/ubuntu/play/picvt/blog/img/cmd/cmd02.png)'
> 
> （PS：感觉问题好多，我只测试过自己遇到过的case，所以有好多情况可能不适配 😭 抱歉抱歉~~）

嗯嗯呢，但是对代码bug有一种执迷不悟😭，今天晚上又过了一遍错误，最暴力的想法是加个replace替换『)』?但是心里不够舒坦
方法一，换个正则，`picvt_local.py`，第九行
```
self.img_pattern = re.compile(r'!\[(.+?)\)', re.M)
```
方法二，简化代码，沿用方法一的正则，去掉`url_pattern`的规则
```
    def __init__(self, params=None):
        self.img_pattern = re.compile ( r'!\[(.+?)\)', re.M )

    def extract(self, content, params):
        url_list = []
        result = self.img_pattern.findall(content)
        for url in result:
            url = url.split("](")[1]
            if url:
                url_list.append(url)
        return url_list
```
已经测试过可以使用了，而且这个正则可以匹配到除了`/statics/text.png`还有`../statics/text.png`，我用了好多相对路径之前不支持😭
另外，学长提交后，我会在我的博客引用一下你的仓库😁，[Lsky图床迁入方案 | 张时贰](https://zhangshier.vip/posts/45734/)，今天太晚了，刚才忙别的去了还没加进去，明天加一下

**Comment by caibingcheng** [comment]
> @GC-ZF 

方法二切割方法不可行：
1. `![xxx](url xxx)` 这种常见case无法匹配、尾部空格也无法剔除

（PS：使用相对路径实际上可能会有些问题，这大部分时候意味着文章的url和statics url也要保持相对路径的结构，但是并不能保证所有case满足这种条件）

我昨天修改了一下local url的正则，可以匹配`![xxx](url)和![xxx](url xxx)`的情况，忘记提交了。。。今天晚点提交一下。

**Comment by GC-ZF** [comment]
> 方法二切割方法不可行：
> 1. `![xxx](url xxx)` 这种常见case无法匹配、尾部空格也无法剔除
> 
> （PS：使用相对路径实际上可能会有些问题，这大部分时候意味着文章的url和statics url也要保持相对路径的结构，但是并不能保证所有case满足这种条件）
> 
> 我昨天修改了一下local url的正则，可以匹配`![xxx](url)和![xxx](url xxx)`的情况，忘记提交了。。。今天晚点提交一下。

确实确实，不好意思，当时只验证了路径获取对不对，没考虑到多种情况
